exonum/src/encoding/fields.rs

```diff
@@ -258,6 +258,64 @@ impl<'a> Field<'a> for DateTime<Utc> {
     }
 }
 
+impl<'a> Field<'a> for Duration {
+    fn field_size() -> Offset {
+        (mem::size_of::<i64>() + mem::size_of::<i32>()) as Offset
+    }
+
+    unsafe fn read(buffer: &'a [u8], from: Offset, to: Offset) -> Self {
+        let secs = LittleEndian::read_i64(&buffer[from as usize..from as usize + 8]);
+        let nanos = LittleEndian::read_i32(&buffer[from as usize + 8..to as usize]);
+
+        // Assuming that buffer was checked and Duration object can be constructed.
+        Duration::seconds(secs) + Duration::nanoseconds(i64::from(nanos))
+    }
+
+    fn write(&self, buffer: &mut Vec<u8>, from: Offset, to: Offset) {
+        let secs = self.num_seconds();
+        let nanos_as_duration = *self - Duration::seconds(secs);
+        // Since we're working with only nanos, no overflow is expected here.
+        let nanos = nanos_as_duration.num_nanoseconds().unwrap() as i32;
+        LittleEndian::write_i64(
+            &mut buffer[from as usize..to as usize - mem::size_of_val(&nanos)],
+            secs,
+        );
+        LittleEndian::write_i32(
+            &mut buffer[from as usize + mem::size_of_val(&secs)..to as usize],
+            nanos,
+        );
+    }
+
+    fn check(
+        buffer: &'a [u8],
+        from: CheckedOffset,
+        to: CheckedOffset,
+        latest_segment: CheckedOffset,
+    ) -> Result {
+        debug_assert_eq!((to - from)?.unchecked_offset(), Self::field_size());
+        let from_unchecked = from.unchecked_offset() as usize;
+        let to_unchecked = to.unchecked_offset() as usize;
+
+        let secs = LittleEndian::read_i64(&buffer[from_unchecked..from_unchecked + 8]);
+        let nanos = LittleEndian::read_i32(&buffer[from_unchecked + 8..to_unchecked]);
```

Comment:
Can `|nanos| > 10 ** 9` lead to some problems? For one, it seems that these values would fail a roundtrip from/to the serialized form. The same question about positive / negative `nanos` - they can lead to multiple presentations of the same duration (e.g., `4 s + 4e8 ns = 5 s - 6e8 ns`).

Reply:
According to `chrono` sources negative nanos are impossible:
```rust
pub struct Duration {
    secs: i64,
    nanos: i32, // Always 0 <= nanos < NANOS_PER_SEC
}
```

Reply:
Yeah, but I don't see how it is reflected in `read` or `write` (unless I'm mistaken, `write` actually produces non-positive `nanos` for negative durations; e.g., `-5.4s` would be written as `{ secs: -5, nanos: -400_000_000 }`).

Reply:
Well, you're right. But since |nanos| does not exceed 10 ** 9, I see no problem here:
object is splitted into seconds and nanoseconds, and then constructed back.
Handling of this process is up to Duration object, since it was decided to completely encapsulate internals (even `nanos_mod_sec` method is private).
The only solution I see is ещ completely repeat library actions, but this way is pretty much error-prone, becuase internals may change.
Should we really strictly follow internal implementation when authors don't want us to base on it?

Reply:
I see a potential problem in that a single `Duration` can be represented in multiple ways (i.e., not survive a roundtrip); I've mentioned an example in another comment. If there was a well-specified canonical presentation of `Duration` enforced through `check` and adhered to by `write`, I would not care which one it was (e.g., non-positive `nanos` for negative `secs` would do just fine).